xapian-letor/ranker/ranker.cc

```diff
@@ -192,35 +192,31 @@ write_to_file(const std::vector<Xapian::FeatureVector> & list_fvecs, const strin
     }
 }
 
+// Query file is in the format: <qid> '<query_string>'
 static std::pair<string, string>
 parse_query_string(const string & query_line, int line_number)
 {
-    vector<string> token;
-    size_t i = 0;
-    size_t j = query_line.find_first_of(' ', i);
-    token.push_back(query_line.substr(i, j - i));
+    string::size_type i = 0;
+    string::size_type j = query_line.find_first_of(' ', i);
+    string qid = query_line.substr(i, j - i);
+    string querystr;
     if (j != string::npos) {
-	i = query_line.find_first_of("'", j);
-	j = string::npos;
-	if (j > i) {
-	    token.push_back(query_line.substr(i, j - i));
+	i = query_line.find_first_not_of("' ", j);
+	j = query_line.length() - 1;
+	// check if the last character is '
+	if (query_line[j] != '\'') {
+	    throw LetorParseError("Could not parse Query file at line:" +
+				   str(line_number));
 	}
+	querystr = query_line.substr(i, j - i);
+    } else {
+	throw LetorParseError("Could not parse Query file at line:" +
+			       str(line_number));
```

Comment:
This exception could be thrown earlier, right after we determine the value of `j`.

Saving a bit of work in an exceptional case generally isn't important (because such cases should happen rarely, and exception handling is slow anyway), but dealing with the exceptional cases promptly often simplifies things because there are then fewer states to consider in the code which follows. E.g. here we wouldn't have to think about what `query_line.substr(i, j - i)` does when `j` is `string::npos` if we dealt with that case right away.

Reply:
fixed in a fixup commit.